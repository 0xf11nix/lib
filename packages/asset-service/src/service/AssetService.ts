import { CAIP2, CAIP19, caip19, WellKnownChain } from '@shapeshiftoss/caip'
import { Asset, AssetDataSource, BaseAsset, OmittedTokenAssetFields } from '@shapeshiftoss/types'
import axios from 'axios'

import assetsDescriptions from './descriptions.json'
import { getRenderedIdenticonBase64, IdenticonOptions } from './GenerateAssetIcon'
import localAssetData from './generatedAssetData.json'

export const flattenAssetData = (assetData: BaseAsset[]): Asset[] => {
  const flatAssetData: Asset[] = []
  for (const baseAsset of assetData) {
    const newAsset = { ...baseAsset }
    delete newAsset.tokens
    flatAssetData.push(newAsset)
    if (baseAsset.tokens) {
      for (const tokenAsset of baseAsset.tokens) {
        const omittedFields: OmittedTokenAssetFields = {
          slip44: baseAsset.slip44,
          explorer: baseAsset.explorer,
          explorerAddressLink: baseAsset.explorerAddressLink,
          explorerTxLink: baseAsset.explorerTxLink
        }
        flatAssetData.push({
          ...tokenAsset,
          ...omittedFields
        })
      }
    }
  }
  return flatAssetData
}

export const indexAssetData = (flatAssetData: Asset[]): IndexedAssetData => {
  return flatAssetData.reduce<IndexedAssetData>(
    (acc, val) => {
      acc.byAssetId.set(val.assetId, val)
      const { chainId } = caip19.fromCAIP19(val.assetId)
      const byChainId = (() => {
        let out = acc.byChainId.get(chainId)
        if (!out) {
          out = new Set()
          acc.byChainId.set(chainId, out)
        }
        return out
      })()
      byChainId.add(val)
      return acc
    },
    {
      byChainId: new Map(),
      byAssetId: new Map()
    }
  )
}

export type IndexedAssetData = {
  byAssetId: Map<CAIP19, Asset>
  byChainId: Map<CAIP2, Set<Asset>>
}

type DescriptionData = Readonly<{
  description: string
  isTrusted?: boolean
}>

export class AssetService {
  private assetFileUrl?: string

  private initialized = false
  private assetData: BaseAsset[]
  private indexedAssetData: IndexedAssetData

  constructor(assetFileUrl?: string) {
    this.assetFileUrl = assetFileUrl
  }

  private checkInitialized() {
    if (!this.initialized) throw new Error('Asset service not initialized')
  }

  /**
   * Get asset data from assetFileUrl and flatten it for easier use
   */
  async initialize() {
    if (this.initialized) return

    try {
      if (!this.assetFileUrl) throw new Error('No assetFileUrl')
      const { data } = await axios.get<BaseAsset[]>(this.assetFileUrl)
      if (!Array.isArray(data)) {
        throw new Error(`Asset Initialize: Return value ${data} is not valid`)
      }
      this.assetData = data
    } catch (err) {
      this.assetData = localAssetData as unknown as BaseAsset[]
    }

    this.indexedAssetData = indexAssetData(flattenAssetData(this.assetData))
    this.initialized = true
  }

  async byChainId(chainId: CAIP2): Promise<Set<Asset>> {
    const out = this.indexedAssetData.byChainId.get(chainId)
    if (!out) return new Set()
    return out
  }

  async byAssetId(assetId: CAIP19): Promise<Asset> {
    const out = this.indexedAssetData.byAssetId.get(assetId)
    if (!out) throw new Error(`no asset data matching assetId ${assetId}`)
    return out
  }

  async description({ asset }: { asset: Asset }): Promise<DescriptionData> {
    const descriptions: Record<string, string> = assetsDescriptions

    // Return overridden asset description if it exists and add isTrusted for description links
    if (descriptions[asset.assetId]) {
      return {
        description: descriptions[asset.assetId],
        isTrusted: true
      }
    }

    if (asset.dataSource !== AssetDataSource.CoinGecko) {
      return { description: '' }
    }

    const { chainId, assetReference: tokenId } = caip19.fromCAIP19(asset.assetId)
    const contractUrl = typeof tokenId === 'string' ? `/contract/${tokenId}` : ''

    const coingeckoCoinId = (() => {
      switch (chainId) {
        case WellKnownChain.BitcoinMainnet:
        case WellKnownChain.BitcoinTestnet:
          return 'bitcoin'
        case WellKnownChain.EthereumMainnet:
        case WellKnownChain.EthereumRopsten:
        case WellKnownChain.EthereumRinkeby:
        case WellKnownChain.EthereumKovan:
          return 'ethereum'
        case WellKnownChain.CosmosHubMainnet:
        case WellKnownChain.CosmosHubVega:
          return 'cosmos'
        case WellKnownChain.OsmosisMainnet:
        case WellKnownChain.OsmosisTestnet:
          return 'osmosis'
        default:
          throw new Error(`coingecko doesn't support chainId ${chainId}`)
      }
    })()

    try {
      type CoinData = {
        description: {
          en: string
        }
      }
      const { data } = await axios.get<CoinData>(
        `https://api.coingecko.com/api/v3/coins/${coingeckoCoinId}${contractUrl}`
      )

      return { description: data?.description?.en ?? '' }
    } catch (e) {
      throw new Error(
        `AssetService:description: no description available for ${tokenId} on chain ${coingeckoCoinId}`
      )
    }
  }

  async generateAssetIconBase64(
    identity: string,
    text?: string,
    options?: IdenticonOptions
  ): Promise<string> {
    return getRenderedIdenticonBase64(identity, text, options)
  }
}
