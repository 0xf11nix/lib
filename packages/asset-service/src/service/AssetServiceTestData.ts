import { Asset, AssetDataSource, BaseAsset } from '@shapeshiftoss/types'

export const ETHMockedAsset: BaseAsset = {
  assetId: 'eip155:1/slip44:60',
  dataSource: AssetDataSource.CoinGecko,
  symbol: 'ETH',
  name: 'Ethereum',
  precision: 18,
  slip44: 60,
  color: '#FFFFFF',
  secondaryColor: '#FFFFFF',
  icon: 'https://assets.coincap.io/assets/icons/eth@2x.png',
  explorer: 'https://etherscan.io',
  explorerTxLink: 'https://etherscan.io/tx/',
  explorerAddressLink: 'https://etherscan.io/address/',
  sendSupport: true,
  receiveSupport: true,
  tokens: [
    {
      assetId: 'eip155:1/erc20:0x7fc66500c84a76ad7e9c93437bfc5ac33e2ddae9',
      name: 'Aave',
      precision: 18,
      color: '#FFFFFF',
      dataSource: AssetDataSource.CoinGecko,
      secondaryColor: '#FFFFFF',
      icon: 'https://assets.coingecko.com/coins/images/12645/thumb/AAVE.png?1601374110',
      sendSupport: true,
      receiveSupport: true,
      symbol: 'AAVE'
    },
    {
      assetId: 'eip155:1/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d',
      name: 'Fox',
      precision: 18,
      color: '#FFFFFF',
      dataSource: AssetDataSource.CoinGecko,
      secondaryColor: '#FFFFFF',
      icon: 'https://assets.coincap.io/assets/icons/fox@2x.png',
      sendSupport: true,
      receiveSupport: true,
      symbol: 'FOX'
    }
  ]
}

export const BTCMockedAsset: BaseAsset = {
  assetId: 'bip122:000000000019d6689c085ae165831e93/slip44:0',
  dataSource: AssetDataSource.CoinGecko,
  symbol: 'BTC',
  name: 'Bitcoin',
  precision: 8,
  slip44: 0,
  color: '#FFFFFF',
  secondaryColor: '#FFFFFF',
  icon: 'https://assets.coincap.io/assets/icons/btc@2x.png',
  explorer: 'https://live.blockcypher.com',
  explorerTxLink: 'https://live.blockcypher.com/btc/tx/',
  explorerAddressLink: 'https://live.blockcypher.com/btc/address/',
  sendSupport: false,
  receiveSupport: false
}

export const mockBaseAssets: BaseAsset[] = [
  ETHMockedAsset,
  {
    assetId: 'eip155:3/slip44:60',
    dataSource: AssetDataSource.CoinGecko,
    symbol: 'ETH',
    name: 'Ethereum',
    precision: 18,
    slip44: 60,
    color: '#FFFFFF',
    secondaryColor: '#FFFFFF',
    icon: 'https://assets.coincap.io/assets/icons/eth@2x.png',
    explorer: 'https://etherscan.io',
    explorerTxLink: 'https://etherscan.io/tx/',
    explorerAddressLink: 'https://etherscan.io/address/',
    sendSupport: true,
    receiveSupport: true,
    tokens: [
      {
        assetId: 'eip155:3/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d',
        name: 'Fox',
        precision: 18,
        color: '#FFFFFF',
        secondaryColor: '#FFFFFF',
        dataSource: AssetDataSource.CoinGecko,
        icon: 'https://assets.coincap.io/assets/icons/fox@2x.png',
        sendSupport: true,
        receiveSupport: true,
        symbol: 'FOX'
      }
    ]
  },
  BTCMockedAsset,
  {
    assetId: 'bip122:000000000933ea01ad0ee984209779ba/slip44:0',
    dataSource: AssetDataSource.CoinGecko,
    symbol: 'BTC',
    name: 'Bitcoin',
    precision: 8,
    slip44: 1,
    color: '#FFFFFF',
    secondaryColor: '#FFFFFF',
    icon: 'https://assets.coincap.io/assets/icons/btc@2x.png',
    explorer: 'https://live.blockcypher.com/btc-testnet/',
    explorerTxLink: 'https://live.blockcypher.com/btc-testnet/tx/',
    explorerAddressLink: 'https://live.blockcypher.com/btc-testnet/address/',
    sendSupport: false,
    receiveSupport: false
  }
]

export const mockAssets: Asset[] = [
  {
    assetId: 'eip155:1/slip44:60',
    dataSource: AssetDataSource.CoinGecko,
    symbol: 'ETH',
    name: 'Ethereum',
    precision: 18,
    slip44: 60,
    color: '#FFFFFF',
    secondaryColor: '#FFFFFF',
    icon: 'https://assets.coincap.io/assets/icons/eth@2x.png',
    explorer: 'https://etherscan.io',
    explorerTxLink: 'https://etherscan.io/tx/',
    explorerAddressLink: 'https://etherscan.io/address/',
    sendSupport: true,
    receiveSupport: true
  },
  {
    assetId: 'eip155:1/erc20:0x7fc66500c84a76ad7e9c93437bfc5ac33e2ddae9',
    name: 'Aave',
    precision: 18,
    color: '#FFFFFF',
    dataSource: AssetDataSource.CoinGecko,
    secondaryColor: '#FFFFFF',
    icon: 'https://assets.coingecko.com/coins/images/12645/thumb/AAVE.png?1601374110',
    sendSupport: true,
    receiveSupport: true,
    symbol: 'AAVE',
    slip44: 60,
    explorer: 'https://etherscan.io',
    explorerTxLink: 'https://etherscan.io/tx/',
    explorerAddressLink: 'https://etherscan.io/address/'
  },
  {
    assetId: 'eip155:1/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d',
    name: 'Fox',
    precision: 18,
    dataSource: AssetDataSource.CoinGecko,
    color: '#FFFFFF',
    secondaryColor: '#FFFFFF',
    icon: 'https://assets.coincap.io/assets/icons/fox@2x.png',
    sendSupport: true,
    receiveSupport: true,
    symbol: 'FOX',
    slip44: 60,
    explorer: 'https://etherscan.io',
    explorerTxLink: 'https://etherscan.io/tx/',
    explorerAddressLink: 'https://etherscan.io/address/'
  },
  {
    assetId: 'eip155:3/slip44:60',
    dataSource: AssetDataSource.CoinGecko,
    symbol: 'ETH',
    name: 'Ethereum',
    precision: 18,
    slip44: 60,
    color: '#FFFFFF',
    secondaryColor: '#FFFFFF',
    icon: 'https://assets.coincap.io/assets/icons/eth@2x.png',
    explorer: 'https://etherscan.io',
    explorerTxLink: 'https://etherscan.io/tx/',
    explorerAddressLink: 'https://etherscan.io/address/',
    sendSupport: true,
    receiveSupport: true
  },
  {
    assetId: 'eip155:3/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d',
    name: 'Fox',
    precision: 18,
    dataSource: AssetDataSource.CoinGecko,
    color: '#FFFFFF',
    secondaryColor: '#FFFFFF',
    icon: 'https://assets.coincap.io/assets/icons/fox@2x.png',
    sendSupport: true,
    receiveSupport: true,
    symbol: 'FOX',
    slip44: 60,
    explorer: 'https://etherscan.io',
    explorerTxLink: 'https://etherscan.io/tx/',
    explorerAddressLink: 'https://etherscan.io/address/'
  },
  {
    assetId: 'bip122:000000000019d6689c085ae165831e93/slip44:0',
    dataSource: AssetDataSource.CoinGecko,
    symbol: 'BTC',
    name: 'Bitcoin',
    precision: 8,
    slip44: 0,
    color: '#FFFFFF',
    secondaryColor: '#FFFFFF',
    icon: 'https://assets.coincap.io/assets/icons/btc@2x.png',
    explorer: 'https://live.blockcypher.com',
    explorerTxLink: 'https://live.blockcypher.com/btc/tx/',
    explorerAddressLink: 'https://live.blockcypher.com/btc/address/',
    sendSupport: false,
    receiveSupport: false
  },
  {
    assetId: 'bip122:000000000933ea01ad0ee984209779ba/slip44:0',
    dataSource: AssetDataSource.CoinGecko,
    symbol: 'BTC',
    name: 'Bitcoin',
    precision: 8,
    slip44: 1,
    color: '#FFFFFF',
    secondaryColor: '#FFFFFF',
    icon: 'https://assets.coincap.io/assets/icons/btc@2x.png',
    explorer: 'https://live.blockcypher.com/btc-testnet/',
    explorerTxLink: 'https://live.blockcypher.com/btc-testnet/tx/',
    explorerAddressLink: 'https://live.blockcypher.com/btc-testnet/address/',
    sendSupport: false,
    receiveSupport: false
  }
]
