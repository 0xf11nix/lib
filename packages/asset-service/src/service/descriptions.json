{
  "eip155:1/erc20:0xed4064f376cb8d68f770fb1ff088a3d0f3ff5c4d": "This token represents a liquidity pool on Curve, a leading decentralized exchange. Holders earn fees from users trading in the pool, and can also deposit the LP token on ShapeShift to earn yield. This liquidity pool contains CRV and ETH. To acquire Curve CRV-ETH, deposit CRV and/or ETH on https://curve.fi/crveth/deposit. Please be aware that as liquidity pools are composed of differently-priced assets, they are subject to impermanent loss.",
  "eip155:1/erc20:0x5a6a4d54456819380173272a5e8e9b9904bdf41b": "This token represents a liquidity pool on Curve, a leading decentralized exchange. Holders earn fees from users trading in the pool, and can also deposit the LP token on ShapeShift to earn yield. This liquidity pool contains MIM and 3CRV. 3CRV is another curve LP token that consists of DAI, USDT, and USDC. To acquire crvMIM, deposit MIM, DAI, USDT, and/or USDC on [Curve](https://curve.fi/mim/deposit). Please be aware that as liquidity pools are composed of differently-priced assets, they are subject to impermanent loss but because all of the assets in the pool are pegged to $1, impermanent loss is unlikely and would only result if one of the assets in the pool lost its peg.",
  "eip155:1/erc20:0x55a8a39bc9694714e2874c1ce77aa1e599461e18": "This token represents a liquidity pool on Curve, a leading decentralized exchange. Holders earn fees from users trading in the pool, and can also deposit the LP token on [ShapeShift](/defi/earn) to earn yield. This liquidity pool contains [MIM](/assets/eip155:1/erc20:0x99d8a9c45b2eca8864373a26d1459e3dff1e17f3) and [UST](/assets/eip155:1/erc20:0xa47c8bf37f92abed4a126bda807a7b7498661acd). TerraUSD ([UST](/assets/eip155:1/erc20:0xa47c8bf37f92abed4a126bda807a7b7498661acd)) is the decentralized and algorithmic stable coin of the Terra blockchain. To acquire Curve MIM-UST, deposit [MIM](/assets/eip155:1/erc20:0x99d8a9c45b2eca8864373a26d1459e3dff1e17f3) and [UST](/assets/eip155:1/erc20:0xa47c8bf37f92abed4a126bda807a7b7498661acd) on [Curve](https://curve.fi/factory/48/deposit). Please be aware that as liquidity pools are composed of differently-priced assets, they are subject to impermanent loss but because all of the assets in the pool are pegged to $1, impermanent loss is unlikely and would only result if one of the assets in the pool lost its peg."
}


