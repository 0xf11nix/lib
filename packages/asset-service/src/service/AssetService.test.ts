import { AssetNamespace } from '@shapeshiftoss/caip'
import { Asset, AssetDataSource } from '@shapeshiftoss/types'
import axios from 'axios'

import { AssetService, flattenAssetData, indexAssetData } from './AssetService'
import { mockAssets, mockBaseAssets } from './AssetServiceTestData'
import descriptions from './descriptions.json'

jest.mock('axios')

const mockedAxios = axios as jest.Mocked<typeof axios>

const EthAsset: Asset = {
  assetId: 'eip155:3/slip44:60',
  dataSource: AssetDataSource.CoinGecko,
  symbol: 'ETH',
  name: 'Ropsten Testnet Ethereum',
  precision: 18,
  slip44: 1,
  color: '#FFFFFF',
  secondaryColor: '#FFFFFF',
  icon: 'https://assets.coincap.io/assets/icons/eth@2x.png',
  explorer: 'https://ropsten.etherscan.io/',
  explorerTxLink: 'https://ropsten.etherscan.io/tx/',
  explorerAddressLink: 'https://ropsten.etherscan.io/address/',
  sendSupport: false,
  receiveSupport: false
}

jest.mock(
  './descriptions.json',
  () => ({
    'eip155:3/slip44:60': 'overridden description'
  }),
  { virtual: true }
)

describe('AssetService', () => {
  const assetFileUrl = 'http://example.com'
  describe('utilities', () => {
    describe('flattenAssetData', () => {
      it('should flatten data correctly', () => {
        expect(flattenAssetData(mockBaseAssets)).toEqual(mockAssets)
      })
    })
    describe('indexAssetData', () => {
      it('should index data correctly', () => {
        expect(indexAssetData(flattenAssetData(mockBaseAssets))).toMatchSnapshot()
      })
    })
  })

  describe('description', () => {
    it('should return the overridden description if it exists', async () => {
      const assetService = new AssetService(assetFileUrl)

      await expect(assetService.description({ asset: EthAsset })).resolves.toEqual({
        description: 'overridden description',
        isTrusted: true
      })
    })

    it('should return a string if found', async () => {
      const assetDescriptions = descriptions as Record<string, string>
      delete assetDescriptions[EthAsset.assetId]

      const assetService = new AssetService(assetFileUrl)
      const description = { en: 'a blue fox' }
      mockedAxios.get.mockResolvedValue({ data: { description } })
      await expect(assetService.description({ asset: EthAsset })).resolves.toEqual({
        description: description.en
      })
    })

    it('should throw if not found', async () => {
      const assetService = new AssetService(assetFileUrl)
      mockedAxios.get.mockRejectedValue({ data: null })
      const tokenData: Asset = {
        assetId: 'eip155:3/erc20:0x1da00b6fc705f2ce4c25d7e7add25a3cc045e54a',
        explorer: 'https://etherscan.io',
        explorerTxLink: 'https://etherscan.io/tx/',
        explorerAddressLink: 'https://etherscan.io/address/',
        name: 'Test Token',
        precision: 18,
        color: '#FFFFFF',
        dataSource: AssetDataSource.CoinGecko,
        secondaryColor: '#FFFFFF',
        slip44: 60,
        icon: 'https://assets.coingecko.com/coins/images/17049/thumb/BUNNY.png?1626148809',
        sendSupport: true,
        receiveSupport: true,
        symbol: 'TST'
      }
      await expect(
        assetService.description({
          asset: tokenData
        })
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        `"AssetService:description: no description available for 0x1da00b6fc705f2ce4c25d7e7add25a3cc045e54a on chain ethereum"`
      )
    })
  })
})
