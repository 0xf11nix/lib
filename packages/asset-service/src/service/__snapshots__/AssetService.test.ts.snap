// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetService utilities indexAssetData should index data correctly 1`] = `
Object {
  "byAssetId": Map {
    "eip155:1/slip44:60" => Object {
      "assetId": "eip155:1/slip44:60",
      "color": "#FFFFFF",
      "dataSource": "coingecko",
      "explorer": "https://etherscan.io",
      "explorerAddressLink": "https://etherscan.io/address/",
      "explorerTxLink": "https://etherscan.io/tx/",
      "icon": "https://assets.coincap.io/assets/icons/eth@2x.png",
      "name": "Ethereum",
      "precision": 18,
      "receiveSupport": true,
      "secondaryColor": "#FFFFFF",
      "sendSupport": true,
      "slip44": 60,
      "symbol": "ETH",
    },
    "eip155:1/erc20:0x7fc66500c84a76ad7e9c93437bfc5ac33e2ddae9" => Object {
      "assetId": "eip155:1/erc20:0x7fc66500c84a76ad7e9c93437bfc5ac33e2ddae9",
      "color": "#FFFFFF",
      "dataSource": "coingecko",
      "explorer": "https://etherscan.io",
      "explorerAddressLink": "https://etherscan.io/address/",
      "explorerTxLink": "https://etherscan.io/tx/",
      "icon": "https://assets.coingecko.com/coins/images/12645/thumb/AAVE.png?1601374110",
      "name": "Aave",
      "precision": 18,
      "receiveSupport": true,
      "secondaryColor": "#FFFFFF",
      "sendSupport": true,
      "slip44": 60,
      "symbol": "AAVE",
    },
    "eip155:1/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d" => Object {
      "assetId": "eip155:1/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d",
      "color": "#FFFFFF",
      "dataSource": "coingecko",
      "explorer": "https://etherscan.io",
      "explorerAddressLink": "https://etherscan.io/address/",
      "explorerTxLink": "https://etherscan.io/tx/",
      "icon": "https://assets.coincap.io/assets/icons/fox@2x.png",
      "name": "Fox",
      "precision": 18,
      "receiveSupport": true,
      "secondaryColor": "#FFFFFF",
      "sendSupport": true,
      "slip44": 60,
      "symbol": "FOX",
    },
    "eip155:3/slip44:60" => Object {
      "assetId": "eip155:3/slip44:60",
      "color": "#FFFFFF",
      "dataSource": "coingecko",
      "explorer": "https://etherscan.io",
      "explorerAddressLink": "https://etherscan.io/address/",
      "explorerTxLink": "https://etherscan.io/tx/",
      "icon": "https://assets.coincap.io/assets/icons/eth@2x.png",
      "name": "Ethereum",
      "precision": 18,
      "receiveSupport": true,
      "secondaryColor": "#FFFFFF",
      "sendSupport": true,
      "slip44": 60,
      "symbol": "ETH",
    },
    "eip155:3/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d" => Object {
      "assetId": "eip155:3/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d",
      "color": "#FFFFFF",
      "dataSource": "coingecko",
      "explorer": "https://etherscan.io",
      "explorerAddressLink": "https://etherscan.io/address/",
      "explorerTxLink": "https://etherscan.io/tx/",
      "icon": "https://assets.coincap.io/assets/icons/fox@2x.png",
      "name": "Fox",
      "precision": 18,
      "receiveSupport": true,
      "secondaryColor": "#FFFFFF",
      "sendSupport": true,
      "slip44": 60,
      "symbol": "FOX",
    },
    "bip122:000000000019d6689c085ae165831e93/slip44:0" => Object {
      "assetId": "bip122:000000000019d6689c085ae165831e93/slip44:0",
      "color": "#FFFFFF",
      "dataSource": "coingecko",
      "explorer": "https://live.blockcypher.com",
      "explorerAddressLink": "https://live.blockcypher.com/btc/address/",
      "explorerTxLink": "https://live.blockcypher.com/btc/tx/",
      "icon": "https://assets.coincap.io/assets/icons/btc@2x.png",
      "name": "Bitcoin",
      "precision": 8,
      "receiveSupport": false,
      "secondaryColor": "#FFFFFF",
      "sendSupport": false,
      "slip44": 0,
      "symbol": "BTC",
    },
    "bip122:000000000933ea01ad0ee984209779ba/slip44:0" => Object {
      "assetId": "bip122:000000000933ea01ad0ee984209779ba/slip44:0",
      "color": "#FFFFFF",
      "dataSource": "coingecko",
      "explorer": "https://live.blockcypher.com/btc-testnet/",
      "explorerAddressLink": "https://live.blockcypher.com/btc-testnet/address/",
      "explorerTxLink": "https://live.blockcypher.com/btc-testnet/tx/",
      "icon": "https://assets.coincap.io/assets/icons/btc@2x.png",
      "name": "Bitcoin",
      "precision": 8,
      "receiveSupport": false,
      "secondaryColor": "#FFFFFF",
      "sendSupport": false,
      "slip44": 1,
      "symbol": "BTC",
    },
  },
  "byChainId": Map {
    "eip155:1" => Set {
      Object {
        "assetId": "eip155:1/slip44:60",
        "color": "#FFFFFF",
        "dataSource": "coingecko",
        "explorer": "https://etherscan.io",
        "explorerAddressLink": "https://etherscan.io/address/",
        "explorerTxLink": "https://etherscan.io/tx/",
        "icon": "https://assets.coincap.io/assets/icons/eth@2x.png",
        "name": "Ethereum",
        "precision": 18,
        "receiveSupport": true,
        "secondaryColor": "#FFFFFF",
        "sendSupport": true,
        "slip44": 60,
        "symbol": "ETH",
      },
      Object {
        "assetId": "eip155:1/erc20:0x7fc66500c84a76ad7e9c93437bfc5ac33e2ddae9",
        "color": "#FFFFFF",
        "dataSource": "coingecko",
        "explorer": "https://etherscan.io",
        "explorerAddressLink": "https://etherscan.io/address/",
        "explorerTxLink": "https://etherscan.io/tx/",
        "icon": "https://assets.coingecko.com/coins/images/12645/thumb/AAVE.png?1601374110",
        "name": "Aave",
        "precision": 18,
        "receiveSupport": true,
        "secondaryColor": "#FFFFFF",
        "sendSupport": true,
        "slip44": 60,
        "symbol": "AAVE",
      },
      Object {
        "assetId": "eip155:1/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d",
        "color": "#FFFFFF",
        "dataSource": "coingecko",
        "explorer": "https://etherscan.io",
        "explorerAddressLink": "https://etherscan.io/address/",
        "explorerTxLink": "https://etherscan.io/tx/",
        "icon": "https://assets.coincap.io/assets/icons/fox@2x.png",
        "name": "Fox",
        "precision": 18,
        "receiveSupport": true,
        "secondaryColor": "#FFFFFF",
        "sendSupport": true,
        "slip44": 60,
        "symbol": "FOX",
      },
    },
    "eip155:3" => Set {
      Object {
        "assetId": "eip155:3/slip44:60",
        "color": "#FFFFFF",
        "dataSource": "coingecko",
        "explorer": "https://etherscan.io",
        "explorerAddressLink": "https://etherscan.io/address/",
        "explorerTxLink": "https://etherscan.io/tx/",
        "icon": "https://assets.coincap.io/assets/icons/eth@2x.png",
        "name": "Ethereum",
        "precision": 18,
        "receiveSupport": true,
        "secondaryColor": "#FFFFFF",
        "sendSupport": true,
        "slip44": 60,
        "symbol": "ETH",
      },
      Object {
        "assetId": "eip155:3/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d",
        "color": "#FFFFFF",
        "dataSource": "coingecko",
        "explorer": "https://etherscan.io",
        "explorerAddressLink": "https://etherscan.io/address/",
        "explorerTxLink": "https://etherscan.io/tx/",
        "icon": "https://assets.coincap.io/assets/icons/fox@2x.png",
        "name": "Fox",
        "precision": 18,
        "receiveSupport": true,
        "secondaryColor": "#FFFFFF",
        "sendSupport": true,
        "slip44": 60,
        "symbol": "FOX",
      },
    },
    "bip122:000000000019d6689c085ae165831e93" => Set {
      Object {
        "assetId": "bip122:000000000019d6689c085ae165831e93/slip44:0",
        "color": "#FFFFFF",
        "dataSource": "coingecko",
        "explorer": "https://live.blockcypher.com",
        "explorerAddressLink": "https://live.blockcypher.com/btc/address/",
        "explorerTxLink": "https://live.blockcypher.com/btc/tx/",
        "icon": "https://assets.coincap.io/assets/icons/btc@2x.png",
        "name": "Bitcoin",
        "precision": 8,
        "receiveSupport": false,
        "secondaryColor": "#FFFFFF",
        "sendSupport": false,
        "slip44": 0,
        "symbol": "BTC",
      },
    },
    "bip122:000000000933ea01ad0ee984209779ba" => Set {
      Object {
        "assetId": "bip122:000000000933ea01ad0ee984209779ba/slip44:0",
        "color": "#FFFFFF",
        "dataSource": "coingecko",
        "explorer": "https://live.blockcypher.com/btc-testnet/",
        "explorerAddressLink": "https://live.blockcypher.com/btc-testnet/address/",
        "explorerTxLink": "https://live.blockcypher.com/btc-testnet/tx/",
        "icon": "https://assets.coincap.io/assets/icons/btc@2x.png",
        "name": "Bitcoin",
        "precision": 8,
        "receiveSupport": false,
        "secondaryColor": "#FFFFFF",
        "sendSupport": false,
        "slip44": 1,
        "symbol": "BTC",
      },
    },
  },
}
`;
